#' Wrapper function for annotatr annotations
#'
#' Annotate the result of \code{methylSigCalc} to \code{annotatr} annotations, while also categorizing regions tested as hyper-methylated in either group or "No DM".
#'
#' @section Annotation structure:
#' Annotations from \code{annotatr} are embedded in the \code{myDiff} \code{GRanges} object as a column of \code{GRanges} (named \code{annot}). Of note is that a region annotated to multiple annotations will appear in as many rows, one for each annotation. A convenience column, \code{locus_id}, clearly indicates when a locus is multiply annotated. The easiest way to see all tested loci (CpGs or regions) and their annotations as a flat table is to coerce the result with \code{as.data.frame}. The \code{mcols} of \code{annot} are:
#' \describe{
#'   \item{id}{ A unique ID for the annotation. }
#'   \item{tx_id}{ Either a UCSC knownGene transcript ID (genic annotations) or a Ensembl transcript ID (lncRNA annotations), }
#'   \item{gene_id}{ Entrez ID. }
#'   \item{symbol}{ The gene symbol from the \code{org.*.eg.db} mapping from the Entrez ID. }
#'   \item{type}{ A code of the form \code{[genome]_[type]_[name]} indicating the annotation type. }
#' }
#'
#' @param myDiff A \code{GRanges} object resulting from \code{methylSigCalc} that contains all CpG sites that are tested for differential methylation.
#' @param dmcList A \code{logical} of the same length as \code{myDiff} defining the DMCs or DMRs.
#' @param annotations A \code{GRanges} object resulting from \code{annotatr::build_annotations()} to be used for annotating \code{myDiff} in conjunction with \code{dmcList}. See the documentation for \code{annotatr::build_annotations()} for guidance on how to chose different annotations, or use custom annotations.
#'
#' @return A \code{GRanges} object whose \code{mcols} include all those in \code{myDiff} in addition to:
#' \describe{
#'   \item{locus_ids}{ A unique identifier for the tested locus from \code{methylSigCalc}. }
#'   \item{dm_status}{ A \code{character} indicating which group the CpG / region is hyper-methylated in (based on the \code{levels} of the \code{comparison} column in the \code{pData} used in \code{methylSigCalc}), or "No DM" if it is \code{FALSE} in \code{dmcList}. }
#'   \item{annot}{ A \code{GRanges} containing annotation information for the CpG or region. See the "Annotation structure" section. }
#' }
#'
#' @examples
#' # Annotate the msig_cpgs results
#' data(data, package = 'methylSig')
#'
#' # Use the genome of msig_cpgs and build annotations for CpG features
#' genome = GenomeInfoDb::genome(msig_cpgs)
#' annots = annotatr::build_annotations(genome = genome, annotations = paste(genome, c('cpgs'), sep='_'))
#'
#' # Decide what counts as differentially methylated
#' dmcList = msig_cpgs$fdr < 0.05 & abs(msig_cpgs$meth.diff) > 25
#'
#' myDiff_annotated = methylSigAnnotation(myDiff = msig_cpgs, dmcList = dmcList, annotations = annots)
#'
#' @export
methylSigAnnotation <- function(myDiff, dmcList, annotations) {
    # Create a dm_status column based on dmcList and myDiff$hyper.direction
    myDiff$dm_status = myDiff$hyper.direction
    myDiff$dm_status[!dmcList] = 'No DM'

    # Create a locus_id to enable users to quickly see when a tested locus is annotated to multiple things
    myDiff$locus_id = seq_along(myDiff)

    # Do the annotations
    annotated_dmcs = annotatr::annotate_regions(regions = myDiff, annotations = annotations, ignore.strand = TRUE)

    return(annotated_dmcs)
}
